### PR TITLE
Add FSGroup PSP

### PR DESCRIPTION
**Pull request by EmandM**

### CODE REVIEW

File: Policies/Kubernetes/fsgroup/azurepolicy.json [hunk 1]
```diff
@@ -0,0 +1,79 @@
+{
+  "properties": {
+    "displayName": "Control allowed fsgroup on pods in a Kubernetes Cluster",
```
Comment (RamyasreeChakka) [inline Policies/Kubernetes/fsgroup/azurepolicy.json:3]: nit: FSGroup

File: Policies/Kubernetes/fsgroup/azurepolicy.json [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+    "policyType": "Custom",
+    "mode": "Microsoft.Kubernetes.Data",
+    "description": "This policy controls valid ranges for FsGroup (File System Group) on volumes in a Kubernetes cluster. For instructions on using this policy, please visit https://aka.ms/kubepolicydoc.",
```
Comment (RamyasreeChakka) [inline Policies/Kubernetes/fsgroup/azurepolicy.json:6]: nit: FSGroup

File: Policies/Kubernetes/fsgroup/azurepolicy.json [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+        "metadata": {
+          "displayName": "Namespace exclusions",
+          "description": "List of Kubernetes namespaces to exclude from policy evaluation. Providing a value for this parameter is optional."
+        },
+        "defaultValue": []
+      },
+      "fsgroupRule": {
+        "type": "String",
+        "metadata": {
+          "displayName": "Rule for a pod to confirm to FsGroup ranges.",
```
Comment (RamyasreeChakka) [inline Policies/Kubernetes/fsgroup/azurepolicy.json:36]: nit: FSGroup as it is File System Group. Comment applicable in other places as well where 'FsGroup' is used.

File: Policies/Kubernetes/fsgroup/azurepolicy.json [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+          "description": "The rule for how pods may confirm to the given FsGroup ranges. Either 'MayRunAs', 'MustRunAs, or 'RunAsAll'. For more information, visit https://kubernetes.io/docs/concepts/policy/pod-security-policy/#volumes-and-file-systems"
+        }
+      },
+      "validFsgroupRanges": {
+        "type": "Object",
+        "metadata": {
+          "displayName": "Valid ranges for Pod FsGroup",
+          "description": "Enter valid ranges from 'min' to max' for the fsgroup used in a pod's security context. Format: { \"ranges\": {\"min\": <minimum valid range>, \"max\": <maximum valid range>\" }"
```
Comment (RamyasreeChakka) [inline Policies/Kubernetes/fsgroup/azurepolicy.json:44]: I think we can remove format and give an example. By the way array is missing in the format you specified.

File: Policies/Kubernetes/fsgroup/azurepolicy.json [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+      },
+      "excludedNamespaces": {
+        "type": "Array",
+        "metadata": {
+          "displayName": "Namespace exclusions",
+          "description": "List of Kubernetes namespaces to exclude from policy evaluation. Providing a value for this parameter is optional."
+        },
+        "defaultValue": []
+      },
+      "fsgroupRule": {
```
Comment (RamyasreeChakka) [inline Policies/Kubernetes/fsgroup/azurepolicy.json:33]: Can you please check if this parameter accepts case-insensitive values during policy assignment? If yes, can you check if REGO honors case-insensitive comparison?

File: Policies/Kubernetes/fsgroup/azurepolicy.json [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+        "type": "String",
+        "metadata": {
+          "displayName": "Rule for a pod to confirm to FSGroup ranges.",
+          "description": "The rule for how pods may confirm to the given FSGroup ranges. Either 'MayRunAs', 'MustRunAs, or 'RunAsAll'. For more information, visit https://kubernetes.io/docs/concepts/policy/pod-security-policy/#volumes-and-file-systems"
```
Comment (RamyasreeChakka) [inline Policies/Kubernetes/fsgroup/azurepolicy.json:37]: nit: use aka.ms link

File: Policies/Kubernetes/fsgroup/rego/src_test.rego [hunk 7]
```diff
@@ -0,0 +1,185 @@
+package k8sazurefsgroup
+
+test_input_fsgroup_allowed_all {
+    input := { "review": input_review_with_fsgroup, "parameters": input_parameters_runasany}
+    results := violation with input as input
+    count(results) == 0
+}
```
Comment (RamyasreeChakka) [inline Policies/Kubernetes/fsgroup/rego/src_test.rego:7]: nit: Add a test combination of input_review_with_securitycontext_no_fsgroup and input_parameters_runasany.

File: Policies/Kubernetes/fsgroup/rego/src_test.rego [hunk 8]
```diff
@@ -0,0 +1,185 @@
...
+test_input_fsgroup_MayRunAs_two_ranges_allowed {
+    input := { "review": input_review_with_fsgroup, "parameters": input_parameters(mayrun, [range_outofrange, range_high_range])}
+    results := violation with input as input
+    count(results) == 0
+}
+test_input_fsgroup_MayRunAs_two_ranges_not_allowed {
+    input := { "review": input_review_with_fsgroup, "parameters": input_parameters(mayrun, [range_outofrange, range_high_range_outofrange])}
+    results := violation with input as input
+    count(results) == 1
```
Comment (RamyasreeChakka) [inline Policies/Kubernetes/fsgroup/rego/src_test.rego:71]: Why is this a violation?
In test_input_fsgroup_MayRunAs_two_ranges_allowed, I see FSGroup value as 1000. Ranges are (1, 1000) and (1200, 2000).
In test_input_fsgroup_MayRunAs_two_ranges_not_allowed, I see FSGroup value as 1000. Ranges are (1, 1000) and (2100, 4000).

In both the above tests, FSGroup value 1000 falls under (1, 1000) bucket and hence violations should not be thrown.

File: Policies/Kubernetes/fsgroup/azurepolicy.json (hunk 5 above)
Comment (EmandM) [reply Policies/Kubernetes/fsgroup/azurepolicy.json:33]: This parameter does not accept case-insensitive values. The Rego does not honor case insensitive comparison.

Comment (RamyasreeChakka) [reply Policies/Kubernetes/fsgroup/azurepolicy.json:33]: Thanks for checking!

**Pull request merged**